satpy/writers/cf_writer.py

```diff
@@ -577,53 +740,345 @@ def update_encoding(dataset, to_netcdf_kwargs, numeric_name_prefix='CHANNEL_'):
+    return dataarray
+
+
+def _collect_cf_dataset(list_dataarrays,
+                        epoch=EPOCH,
+                        flatten_attrs=False,
+                        exclude_attrs=None,
+                        include_lonlats=True,
+                        pretty=False,
+                        include_orig_name=True,
+                        numeric_name_prefix='CHANNEL_'):
+    """Process a list of xr.DataArray and return a dictionary with CF-compliant xr.Dataset.
+
+    Parameters
+    ----------
+    list_dataarrays (list):
+        List of DataArrays to make CF compliant and merge into a xr.Dataset.
+    epoch (str):
+        Reference time for encoding the time coordinates (if available).
+        Example format: "seconds since 1970-01-01 00:00:00".
+        If None, the default reference time is retrieved using `from satpy.cf_writer import EPOCH`
+    flatten_attrs (bool):
+        If True, flatten dict-type attributes.
+    exclude_attrs (list):
+        List of xr.DataArray attribute names to be excluded.
+    include_lonlats (bool):
+        If True, it includes 'latitude' and 'longitude' coordinates also for satpy scene defined on an AreaDefinition.
+        If the 'area' attribute is a SwathDefinition, it always include latitude and longitude coordinates.
+    pretty (bool):
+        Don't modify coordinate names, if possible. Makes the file prettier, but possibly less consistent.
+    include_orig_name (bool).
+        Include the original dataset name as a variable attribute in the xr.Dataset.
+    numeric_name_prefix (str):
+        Prefix to add the each variable with name starting with a digit.
+        Use '' or None to leave this out.
+
+    Returns
+    -------
+    ds : xr.Dataset
+        A partially CF-compliant xr.Dataset
+    """
+    # Create dictionary of input datarrays
+    # --> Since keys=None, it doesn't never retrieve ancillary variables !!!
+    ds_collection = {}
+    for dataarray in list_dataarrays:
+        ds_collection.update(get_extra_ds(dataarray))
+
+    # Check ???
+    got_lonlats = has_projection_coords(ds_collection)
+
+    # Sort dictionary by keys name
+    ds_collection = dict(sorted(ds_collection.items()))
+
+    dict_dataarrays = {}
+    for dataarray in ds_collection.values():
+        dataarray_type = dataarray.dtype
+        if dataarray_type not in CF_DTYPES:
+            warnings.warn(
+                f'dtype {dataarray_type} not compatible with {CF_VERSION}.',
+                stacklevel=3
+            )
+        # Deep copy the datarray since adding/modifying attributes and coordinates
+        dataarray = dataarray.copy(deep=True)
+
+        # Add CF-compliant area information from the pyresample area
+        # - If include_lonlats=True, add latitude and longitude coordinates
+        # - Add grid_mapping attribute to the DataArray
+        # - Return the CRS DataArray as first list element
+        # - Return the CF-compliant input DataArray as second list element
```

Comment:
Please refactor the code so that this workflow is clear

Reply:
Yes